smart_open/compression.py

```diff
@@ -83,9 +86,15 @@ def close_both(*args):
         try:
             outer_close()
         finally:
-            if inner:
-                inner, fp = None, inner
-                fp.close()
+            ex, _, _ = sys.exc_info()
+
+            # https://github.com/RaRe-Technologies/smart_open/issues/684
+            # If exception is raised multipart writer should be terminated to abort file uploading to S3
+            # If not incomplete file will be saved in bucket
+            if type(inner) == MultipartWriter and ex:
```

Comment:
Hi 👋 can this be generalized such that it doesn't just handle S3, but also the other adapters like Azure?

Azure exception handling ref https://github.com/RaRe-Technologies/smart_open/pull/783

Reply:
e.g. by refactoring `tweak_close` to become `tweak_exit`, that patches `.__exit__()` instead of `.close()`?

Reply:
opened https://github.com/RaRe-Technologies/smart_open/pull/786